### PR TITLE
[WIP] Storage version compatibilty mode

### PR DESCRIPTION
**Pull request by alexzielenski**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This change is intended to make the version used for encoding in etcd configurable by a `compatibility-version` flag. The used storage versions will be looked up from a table that takes into account K8s versions each resource was introduced, and version skew. This is intended to make for smoother upgrades.

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:

ATM this is just some basic planning for wiring. Posting as a WIP to source some early feedback and comments about the approach.

/cc @logicalhan @jpbetz  @liggitt 

#### Does this PR introduce a user-facing change?

```release-note
TBD
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/server/storage/resource_encoding_config_compatibility.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+var v1_27 = semver.MustParse("v1.27.0")
+
+func NewCompatibilityResourceConfig() *CompatibilityResourceEncodingConfig {
+	bv := getBinaryVersion()
+	return &CompatibilityResourceEncodingConfig{
+		binaryVersion:        bv,
+		compatibilityVersion: bv,
+		resourceHistory:      ResourceHistory{
+			//!TODO: populate with history
+			// schema.GroupResource{Group: "authorization.k8s.io", Resource: "localsubjectaccessreviews"}: {
```
Comment (alexzielenski) [inline staging/src/k8s.io/apiserver/pkg/server/storage/resource_encoding_config_compatibility.go:42]: The resource history is intended to inform which k8s version a given GVR became supported by k8s. Right now I have it empty. Curious for comments about:

1. The ResourceHistory data structure for tracking when GVRs were supported
2. Reliable ways to keep it populated without manual labor

File: staging/src/k8s.io/apiserver/pkg/server/storage/resource_encoding_config_compatibility.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+			// 	{Version: "v1alpha1", K8sVersion: v1_26},
+			// },
+		},
+		versionSkew: 1,
```
Comment (alexzielenski) [inline staging/src/k8s.io/apiserver/pkg/server/storage/resource_encoding_config_compatibility.go:46]: Configurable version skew to replace the current manual list of overrides

File: pkg/kubeapiserver/default_storage_factory_builder.go [hunk 3]
```diff
@@ -115,12 +114,14 @@ type completedStorageFactoryConfig struct {
...
-	resourceEncodingConfig := resourceconfig.MergeResourceEncodingConfigs(c.DefaultResourceEncoding, c.ResourceEncodingOverrides)
+	// resourceEncodingConfig := resourceconfig.MergeResourceEncodingConfigs(c.DefaultResourceEncoding, c.ResourceEncodingOverrides)
 	storageFactory := serverstorage.NewDefaultStorageFactory(
 		c.StorageConfig,
 		c.DefaultStorageMediaType,
 		c.Serializer,
-		resourceEncodingConfig,
+		// NewCompatibilityResourceConfig is intended to obviate the need for
+		// manual overrides list by taking skew into account.
+		serverstorage.NewCompatibilityResourceConfig(),
```
Comment (alexzielenski) [inline pkg/kubeapiserver/default_storage_factory_builder.go:124]: I think this only affects kubeapiserver types (not aggregator/apiextensions), which override the `EncodeVersioner` for their storages manually

**Review by alexzielenski — COMMENTED** [review]
starting some discussions for review

File: staging/src/k8s.io/apiserver/pkg/server/storage/resource_encoding_config_compatibility.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+}
+
+type ResourceHistory = map[schema.GroupResource][]ResourceSpec
+
+type CompatibilityResourceEncodingConfig struct {
+	binaryVersion        semver.Version
+	compatibilityVersion semver.Version
+	versionSkew          uint8
+	resourceHistory      ResourceHistory
+}
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/storage/resource_encoding_config_compatibility.go:28]: the types could be defined here, but specific k8s versions and history of served types belongs somewhere like pkg/kubeapiserver, etc, not in the generic k8s.io/apiserver library

File: staging/src/k8s.io/apiserver/pkg/server/storage/resource_encoding_config_compatibility.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+	"github.com/blang/semver/v4"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/component-base/version"
+)
+
+type ResourceSpec struct {
+	// Version indicates the Kubernetes version from which this GVR is decodable
+	K8sVersion semver.Version
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/storage/resource_encoding_config_compatibility.go:15]: remember that anything in this package is for generic apiservers, not necessarily kube-apiserver

File: staging/src/k8s.io/apiserver/pkg/server/storage/resource_encoding_config_compatibility.go [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+	// Version indicates the resource version for which support was added
+	Version string
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/storage/resource_encoding_config_compatibility.go:18]: version is somewhat overloaded, is this like "v1beta1", "v1", etc?

Comment (alexzielenski) [reply staging/src/k8s.io/apiserver/pkg/server/storage/resource_encoding_config_compatibility.go:18]: yeah. I thought to use `APIVersion` for this field but to me that usually includes the group too

File: staging/src/k8s.io/apiserver/pkg/server/storage/resource_encoding_config_compatibility.go (hunk 1 above)
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/server/storage/resource_encoding_config_compatibility.go:42]: test/integration/etcd/data.go and staging/src/k8s.io/api/testdata are two places where every group/version/kind we know about in kube-apiserver has an associated entry

We make use of the skew fixture data in staging/src/k8s.io/api/testdata in the etcd test to make sure we don't switch the storage version to a version the previous server did not understand if the previous server understood any non-alpha versions:
https://github.com/kubernetes/kubernetes/blob/master/test/integration/etcd/etcd_storage_path_test.go#L166-L207

I don't know that I'd copy those approaches exactly, but something along those lines could be an idea

File: staging/src/k8s.io/apiserver/pkg/server/storage/resource_encoding_config_compatibility.go (hunk 4 above)
Comment (alexzielenski) [reply staging/src/k8s.io/apiserver/pkg/server/storage/resource_encoding_config_compatibility.go]: Moved all relevant code to pkg/kubeapiserver

**Comment by alexzielenski** [comment]
Attemped to start in a "1.25.0" compatibility mode and got this error...

```
error: failed to start KubeAPI server: failed to create server chain: problem initializing API group "networking.k8s.io" : storage codec doesn't seem to match given type: internal type not encodable: no kind "IPAddress" is registered for version "networking.k8s.io/v1" in scheme "k8s.io/kubernetes/pkg/api/legacyscheme/scheme.go:30"
```

digging into why this occurs...

**Comment by alexzielenski** [comment]
/close


Moving these changes into @siyuanfoundation 's fork:

https://github.com/siyuanfoundation/kubernetes/pull/1

https://github.com/kubernetes/kubernetes/pull/122891

**Pull request closed without merging**